### PR TITLE
v1.10.8-mev0.4

### PR DESCRIPTION
**Pull request by pgebheim**
# Mev-Geth v0.4

Upgrade Steps: https://docs.flashbots.net/flashbots-auction/releases/alpha-v0.4/#upgrade-steps

Changes:
Includes Megabundle support (#85) and a bug fix related to gas double counting (#86)

### CODE REVIEW

File: miner/worker.go [hunk 1]
```diff
@@ -1196,8 +1196,42 @@ func (w *worker) commitNewWork(interrupt *int32, noempty bool, timestamp int64)
 		if w.commitBundle(bundleTxs, w.coinbase, interrupt) {
 			return
 		}
-		w.current.profit.Add(w.current.profit, bundle.totalEth)
+		w.current.profit.Add(w.current.profit, bundle.ethSentToCoinbase)
```
Comment (metachris) [inline miner/worker.go:1199]: Just curious - why has the profic calculation here changed from `totalEth` to `ethSentToCoinbase`?

File: miner/worker.go [hunk 2]
```diff
@@ -1196,8 +1196,42 @@ func (w *worker) commitNewWork(interrupt *int32, noempty bool, timestamp int64)
...
+	if w.flashbots.isMegabundleWorker {
+		megabundle, err := w.eth.TxPool().GetMegabundle(w.flashbots.relayAddr, header.Number, header.Time)
+		log.Info("Starting to process a Megabundle", "relay", w.flashbots.relayAddr, "megabundle", megabundle, "error", err)
+		if err != nil {
+			return // no valid megabundle for this relay, nothing to do
+		}
+		// Flashbots bundle merging duplicates work by simulating TXes and then committing them once more.
+		// Megabundles API focuses on speed and runs everything in one cycle.
+		coinbaseBalanceBefore := w.current.state.GetBalance(w.coinbase)
+		if w.commitBundle(megabundle.Txs, w.coinbase, interrupt) {
```
Comment (metachris) [inline miner/worker.go:1210]: I find it slightly confusing that `commitBundle` returns `true` if it has an error. Would it be better to return an `error`?

File: miner/worker.go [hunk 3]
```diff
@@ -1196,8 +1196,42 @@ func (w *worker) commitNewWork(interrupt *int32, noempty bool, timestamp int64)
...
+				return
+			}
+			if !status && !containsHash(megabundle.RevertingTxHashes, tx.Hash()) {
+				log.Info("Ignoring megabundle because of failing TX", "relay", w.flashbots.relayAddr, "TxHash", tx.Hash())
+				return
+			}
+		}
+		coinbaseBalanceAfter := w.current.state.GetBalance(w.coinbase)
+		coinbaseDelta := big.NewInt(0).Sub(coinbaseBalanceAfter, coinbaseBalanceBefore)
+		w.current.profit = coinbaseDelta
```
Comment (metachris) [inline miner/worker.go:1231]: Would this profit calculation be off if the miner sends (or receives) transactions with value from/to the coinbase address?

File: miner/worker.go (hunk 2 above)
Comment (jparyani) [reply miner/worker.go:1210]: It's intended to follow the same pattern as `commitTransactions`, although I agree it's non-intuitive

**Review by metachris — APPROVED** [review]

**Pull request merged**